src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpointResolver.java

```diff
@@ -377,7 +380,30 @@ static InetSocketAddress getResolvedHttpProxyAddress(@Nonnull String host, int p
         while (targetAddress == null && proxies.hasNext()) {
             Proxy proxy = proxies.next();
             if (proxy.type() == Proxy.Type.DIRECT) {
-                break;
+                // Proxy.NO_PROXY with a DIRECT type is returned in two cases:
+                // - when no proxy (none) has been configured in the JVM (either with system properties or by the operating system)
+                // - when the URI host is part of the exclusion list defined by system property -Dhttp.nonProxyHosts
+                //
+                // Unfortunately, the Proxy class does not provide a way to differentiate both cases to fallback to
+                // environment variables only when no proxy has been configured. Therefore, we have to recheck if the URI
+                // host is in the exclusion list.
+                String nonProxyHosts = NetProperties.get("http.nonProxyHosts");
+                if(nonProxyHosts != null && nonProxyHosts.length() != 0) {
+                    RegexpPool exclusionsPool = new RegexpPool();
+                    StringTokenizer stringTokenizer = new StringTokenizer(nonProxyHosts, "|", false);
+                    try {
+                        while(stringTokenizer.hasMoreTokens()) {
+                            exclusionsPool.add(stringTokenizer.nextToken().toLowerCase(), Boolean.TRUE);
```

Comment:
`toLowerCase()` is invoked without locale, it will cause FindBugs failures with #242 I'd guess